File: Classes/Proprietario.java
package Classes;
/* Classe responsável por armazenar um proprietário. Que, por sua vez, possuirá imóveis. */
import java.util.ArrayList;
public class Proprietario {
    private String nome;

    //#############################################################################################################
    // como CPF, Identidade e CEP não serão parte de nenhuma operação matemática, considerei criá-los como variáveis do tipo String.
    // Além disso, todos podem começar com um "0" a esquerda, o que não seria possível para um tipo Int
    //#############################################################################################################


    private String cpf;
    private String identidade;
    private Endereco endereco;
    private ArrayList<Imovel> imoveis;


    public Proprietario(String nome, String cpf, String identidade, String cep, String estado, String cidade, String rua, int numero) {
        this.nome = nome;
        this.cpf = cpf;
        this.identidade = identidade;
        this.endereco = new Endereco(cep, estado, cidade, rua, numero);
        this.imoveis = new ArrayList<Imovel>();

    }


    /*
    Acredito que não exista a possibilidade de um objeto proprietário atualizar atributos como nome, cpf e identidade, uma vez que esses dados são fixos.
    E a atualização implica um novo proprietário, ou seja, ium novo objeto.
     */
    public String getNome() {
        return nome;
    }

    public String getCpf() {
        return cpf;
    }

    public String getIdentidade() {
        return identidade;
    }


    // Ainda não tenho certeza se seria melhor criar o objeto imóvel fora da chamada do addImovel, ou se é melhor criar o objeto dentro da chamada do addImovel.
    // De qualquer modo, tal implementação subentende que o imóvel não irá existir sem um proprietário(dentro do programa).
    public void addImovel(Imovel imovel) {
        this.imoveis.add(imovel);
    }

    //#############################################################################################################
    /*
    TALVEZ UMA ESTRUTURA DESSE TIPO TENHA UTILIDADE DEPOIS. FICA AQUI COMO REFERÊNCIA PARA O FUTURO.

    public void getEndereco() {
        System.out.println(this.endereco.getCep());
        System.out.println(this.endereco.getEstado());
        System.out.println(this.endereco.getCidade());
        System.out.println(this.endereco.getRua());
        System.out.println(this.endereco.getNumero());
    }
    //#############################################################################################################
*/

    // Possibilidade de atualização de endereço provavelmente será requisitada mais pra frente. Então, criei um método para isso.
    public void setEndereco(String cep, String estado, String cidade, String rua, int numero) {
        this.endereco.setEndereco(cep, estado, cidade, rua, numero);
    }
    /* Para que essa sobrecarga funcione, encontrar uma boa solução para o Enum.

    public void setEndereco(String cep, String rua, int numero) {
        this.endereco.setEndereco(cep, this.endereco.getEstado(), this.endereco.getCidade(), rua, numero);
    }
    */


    public Endereco getEndereco() {
        return endereco;
    }

    public void listaImoveis(String tipo) {
        int cnt = 1;
        System.out.println("\nImóveis do tipo " + tipo + " do proprietário " + this.nome + ":\n");
        System.out.println("--------------------------------------------------------------------------------------");
        for (Imovel imovel : imoveis) {
            if (imovel.getTipo().equals(tipo)) {
                System.out.println("Imóvel " + cnt++ + ":");
                System.out.println("Classes.Tipo: " + imovel.getTipo());
                System.out.println("CEP: " + imovel.getEndereco().getCep());
                System.out.println("Estado: " + imovel.getEndereco().getEstado());
                System.out.println("Cidade: " + imovel.getEndereco().getCidade());
                System.out.println("Rua: " + imovel.getEndereco().getRua());
                System.out.println("Número: " + imovel.getEndereco().getNumero());
                System.out.println("Utilidade: " + imovel.getUtilidade());
                System.out.println("IPTU: " + imovel.getIptu() + "\n");
                System.out.println("#################################################################################\n");
            }
        }
    }

    public ArrayList<Imovel> getImoveis() {
        return imoveis;
    }

/*
    public static void main(String[] args) {
        Classes.Proprietario proprietario = new Classes.Proprietario("João", "12345678901", "123456789", "12345678", "SP", "São Paulo", "Rua dos Bobos", 0);
        proprietario.addImovel(new Classes.Imovel("12345578", "SZ", "São Paulo", "Rua dos Bobos", 0, "Casa", "Estadia", 123.50F));
        proprietario.addImovel(new Classes.Imovel("12345578", "Rua dos Bobos", 0, "Casa", "Estadia", 123.50F));
        proprietario.addImovel(new Classes.Imovel("12345578", "SP", "São Paulo", "Rua dos Bobos", 0, "Apartamento", "Estadia", 123.50F));
        proprietario.listaImoveis("Casa");
    }
    */
}

File: incremental_python/proprietario.py
# classe de proprietário questão 1 letra b
import endereco
import imovel

class proprietario:

    def __init__(self, nome, cpf, identidade, cep, estado, cidade, rua, numero):
        self.nome = nome
        self.cpf = cpf
        self.identidade = identidade
        self.endereco = endereco.endereco(cep, estado, cidade, rua, numero)
        self.ListaImoveis = []

    # falta a sobrecarga do construtor

    #gets e sets

    def get_nome(self):
        return self.nome
    
    def set_nome(self, nome):
        self.nome = nome

    def get_cpf(self):
        return self.cpf
    
    def set_cpf(self, cpf):
        self.cpf = cpf

    def get_identidade(self):
        return self.identidade
    
    def set_identidade(self, identidade):
        self.identidade = identidade

    def get_cep(self):
        return self.endereco.get_cep()
    
    def set_cep(self, cep):
        self.endereco.set_cep(cep)

    def get_estado(self):
        return self.endereco.get_estado()
    
    def set_estado(self, estado):
        self.endereco.set_estado(estado)

    def get_cidade(self):
        return self.endereco.get_cidade()
    
    def set_cidade(self, cidade):
        self.endereco.set_cidade(cidade)

    def get_rua(self):
        return self.endereco.get_rua()
    
    def set_rua(self, rua):
        self.endereco.set_rua(rua)

    def get_numero(self):
        return self.endereco.get_numero()
    
    def set_numero(self, numero):
        self.endereco.set_numero(numero)

    def get_iptu(self):
        return self.endereco.get_iptu()
    
    def set_iptu(self, iptu):
        self.endereco.set_iptu(iptu)

    def get_listaImoveis(self):
        return self.ListaImoveis
    
    def get_endereco(self):
        return self.endereco

    #questão 1 letra d (ATUALIZA ENDEREÇO)

    def atualizaEndereco(self, cep, rua, numero, estado=None, cidade=None):
        self.endereco.set_cep(cep)
        if estado is not None or cidade is not None:
            self.endereco.set_estado(estado)
            self.endereco.set_cidade(cidade)
        self.rua = rua
        self.numero = numero

    #sobrecarga de método NÃO FUNCIONA

#   def atualizaEndereco(self, cep, rua, numero):
#        self.cep = cep
#        self.rua = rua
#        self.numero = numero

    #questão 2 cadastra imovel

    def cadastraImovel(self, imovel):
        if imovel.get_endereco().get_cep() == self.endereco.get_cep() and imovel.get_endereco().get_rua() == self.endereco.get_rua() and imovel.get_endereco().get_numero() == self.endereco.get_numero():
            print("\033[31mVocê não pode cadastrar um imóvel no mesmo endereço do proprietário\033[0m")
            return False
        elif(imovel in self.ListaImoveis):
            print("Imóvel já cadastrado")
            return False
        else:
            self.ListaImoveis.append(imovel)
            print("Imóvel cadastrado com sucesso")
            return True
        
    #questão 2 letra a (LISTA IMOVEIS)

    def listaImoveis(self, tipo):
        tipo = tipo.lower()
        tipo = tipo.capitalize()
        if tipo != "Todos":
            cnt = 1
            print("Listando imóveis do tipo " + tipo + ": \n")
            for i in self.get_listaImoveis():
                if i.get_tipo() == tipo:
                    print("Imóvel " + str(cnt) + ": \n")
                    print(i)
                    cnt += 1
        else:
            cnt = 1
            print("Listando todos os imóveis: \n")
            for i in self.get_listaImoveis():
                print("Imóvel " + str(cnt) + ": \n")
                print(i)
                cnt += 1


    def getImovel(self, index):
        return self.ListaImoveis[index]

    